### PR TITLE
🐛  Normalization correctly propagates deletions to the final tables

### LINKED ISSUE #8830: 🐛 CDC Sources to normalized destinations: deleted_at field is not getting updated for deleted records.
Is this your first time deploying Airbyte: Yes
OS Version / Instance: Compute engine n2-highmem-8
Deployment: Docker
Airbyte Version: 0.33.12-alpha
Source name/version: MySql 0.5.1
Destination name/version: BigQuery 0.5.0

### LINKED ISSUE #11805: CDC Postgres (Incremental | Deduped+history) to BigQuery: deleted_at field is not getting updated for deleted records.
## Environment
* **Airbyte version**:  `0.35.56-alpha`
* **OS Version / Instance**: GCP
* **Deployment**: Docker
* **Source Connector and version**: postgres `0.4.9`
* **Destination Connector and version**: BigQuery
* **Severity**: High
* **Step where error happened**: Sync job

## Current Behavior
We’re observing the same replication problem as - [:bug: CDC MySQL (Deduped+history) to Bigquery: deleted_at field is not getting updated for deleted records. · Issue #8830 · airbytehq/airbyte · GitHub](https://github.com/airbytehq/airbyte/issues/8830), for postgres, where if a row is deleted in source, the destination continues to display the row in the replicated table. 

The <table_name>_scd shows _is_airbyte_active column which reflects the deletion, which indicates that there’s a bug in the transformation logic that builds the main table.

## Expected Behavior
Sync job should populate the `_deleted_at` field for syncing when syncing a row which has been deleted at the source.

## Are you willing to submit a PR?
Yes. This is a blocker for us and would be happy to open a PR, if pointed in the right direction.

┆Issue is synchronized with this [Monday item](https://airbyte-bunch.monday.com/boards/2536787439/pulses/2537092049) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by edgao**
## What
Closes https://github.com/airbytehq/airbyte/issues/8830

## How
Adds a post hook on the scd model, which deletes potentially-modified data from the final table.

todo: mssql  integration tests (mssql doesn't work at all on m1 right now, need to figure this out)

## Recommended reading order
1. stream_processor.py
2. incremental.sql
3. messages_incremental.txt (added a new record to test for an incremental logic edge case)
4. simple_streams_second_run_row_counts, simple_streams_third_run_row_counts (updated expected record counts to reflect deletions)
5. `normalization_test_output/**`

## 🚨 User Impact 🚨
Final table will no longer contain deleted rows. This is technically a breaking change, but seems fine given that the old behavior was incorrect.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [x] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [x] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

### CODE REVIEW

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 1]
```diff
@@ -1095,6 +1099,11 @@ def add_to_outputs(
             else:
                 # incremental is handled in the SCD SQL already
                 sql = self.add_incremental_clause(sql)
+                if do_deletions:
+                    # TODO figure out how to get SCD table name correctly
+                    config[
+                        "post_hook"
+                    ] = f"""["delete from {{{{ this }}}} where _airbyte_ab_id in ( select _airbyte_ab_id from {{{{ {scd_table} }}}} where _airbyte_active_row = 0 and cast({self.name_transformer.normalize_column_name(self.airbyte_normalized_at)} as {{{{ type_timestamp_with_timezone() }}}}) >= (select max(cast({self.name_transformer.normalize_column_name(self.airbyte_normalized_at)} as {{{{ type_timestamp_with_timezone() }}}})) from {{{{ {scd_table} }}}}))"]"""
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1106]: Is this a post-hook on the final table to delete from scd table that is built before the final table?

that does not seem right or i am not understanding this correctly?

Comment (edgao) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py]: `{{ this }}` is the final table in this context. E.g. from dbt.log:
```sql
with partitioned_data as (
    select row_number() over (
        partition by _airbyte_unique_key
        order by _airbyte_active_row desc, _airbyte_ab_id
    ) as _airbyte_row_num,
    "postgres".test_normalization."dedup_exchange_rate_scd".*
    from "postgres".test_normalization."dedup_exchange_rate_scd"
    where 1=1
    
and coalesce(
    cast(_airbyte_emitted_at as 
    timestamp with time zone
) >= (select max(cast(_airbyte_emitted_at as 
    timestamp with time zone
)) from "postgres".test_normalization."dedup_exchange_rate_scd"),
    
    
    true)

)
delete from "postgres".test_normalization."dedup_exchange_rate" where _airbyte_unique_key in (
    select _airbyte_unique_key from partitioned_data
    where _airbyte_active_row = 0 and _airbyte_row_num = 1
)
```

(i.e. it's deleting from `dedup_exchange_rate` rather than `dedup_exchange_rate_scd`)

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 2]
```diff
@@ -1095,6 +1099,30 @@ def add_to_outputs(
...
+        order by _airbyte_active_row desc, _airbyte_ab_id
+    ) as _airbyte_row_num,
+    {{ scd_table }}.*
+    from {{ scd_table }}
+    where 1=1
+    {{ incremental_clause }}
+)
+delete from {{ '{{ this }}' }} where _airbyte_unique_key in (
+    select _airbyte_unique_key from partitioned_data
+    where _airbyte_active_row = 0 and _airbyte_row_num = 1
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1117]: Why is it necessary to restrict to `_airbyte_row_num = 1`?

Comment (edgao) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1117]: oh yeah, I think adding `and _airbyte_end_at not null {{ incremental_clause }}` to the subquery would make everything work.

the record might have been updated but not deleted, in which case the SCD table would contain a row with active_row=0. But we don't want to delete it from the final table, so the `row_num = 1` is to filter for the most recent SCD entry (which would have active_row=1). E.g. when I create a record and then update it twice, this is the resulting SCD:
```
edgao_test_replicated=> select id, bar, _airbyte_active_row, _ab_cdc_deleted_at from foo_table_scd where id = 12 order by _ab_cdc_updated_at;
-[ RECORD 1 ]-------+---------
id                  | 12
bar                 | test9
_airbyte_active_row | 0
_ab_cdc_deleted_at  |
-[ RECORD 2 ]-------+---------
id                  | 12
bar                 | test9-1
_airbyte_active_row | 0
_ab_cdc_deleted_at  |
-[ RECORD 3 ]-------+---------
id                  | 12
bar                 | test9-2
_airbyte_active_row | 1
_ab_cdc_deleted_at  |
```

(I should probably be adding a lot of comments to this file...)

Comment (edgao) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1117]: hmmm. I actually think the partitioning might be necessary? For handling the situation where a stream emits multiple records with `null` for their cursor field - in this case, the SCD table could contain some records with active_row=0 and end_at=null, even though a later record has active_row=1. ([example](https://github.com/airbytehq/airbyte/blob/cdb972811dec629d232df1dd30e41624eb64ece7/airbyte-integrations/bases/base-normalization/integration_tests/resources/test_simple_streams/data_input/messages.txt#L24-L25))

Does having the incremental clause inside the `partitioned_data` query not prevent scanning the whole SCD table?

File: airbyte-integrations/bases/base-normalization/dbt-project-template/macros/incremental.sql [hunk 3]
```diff
@@ -4,22 +4,22 @@
     - incremental_clause controls the predicate to filter on new data to process incrementally
 #}
 
-{% macro incremental_clause(col_emitted_at)  -%}
-  {{ adapter.dispatch('incremental_clause')(col_emitted_at) }}
+{% macro incremental_clause(col_emitted_at, tablename)  -%}
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/dbt-project-template/macros/incremental.sql:7]: if you add this extra argument of `tablename`, don't you have to add it to all the calls to this macro?

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 4]
```diff
@@ -1038,15 +1040,15 @@ def add_incremental_clause(self, sql_query: str) -> Any:
         template = Template(
             """
 {{ sql_query }}
-{{'{{'}} incremental_clause({{ col_emitted_at }}) {{'}}'}}
+{{ incremental_clause }}
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1043]: what happens when you refer to `incremental_clause` without arguments?

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 5]
```diff
@@ -1038,15 +1040,15 @@ def add_incremental_clause(self, sql_query: str) -> Any:
...
     """
         )
-        sql = template.render(
-            sql_query=sql_query,
-            col_emitted_at=self.get_emitted_at(in_jinja=True),
-        )
+        sql = template.render(sql_query=sql_query, incremental_clause=self.get_incremental_clause("this"))
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1046]: oh ok i see that's where you changed how you call the incremental clause macro

File: airbyte-integrations/bases/base-normalization/dbt-project-template/macros/incremental.sql (hunk 3 above)
Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-normalization/dbt-project-template/macros/incremental.sql:7]: nvm, i am guessing you are indeed changing the calls thanks to https://github.com/airbytehq/airbyte/pull/12846/files#r874512856

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py (hunk 2 above)
Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1117]: On BigQuery for example, the partitions of `*_scd` tables are using `_airbyte_active_row` column for partitions (and some other columns for clustering):
https://github.com/airbytehq/airbyte/blob/cdb972811dec629d232df1dd30e41624eb64ece7/airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/bigquery/test_simple_streams/models/generated/airbyte_incremental/scd/test_normalization/dedup_exchange_rate_scd.sql#L2-L3

So if the where clauses (in the incremental_clause) are referring to these columns, BQ will be able to prune partitions/indexes and avoid full table scans yes.

In conclusion, what I initially thought about:
```
delete from {{ '{{ this }}' }} where _airbyte_unique_key in ( select _airbyte_unique_key from {{ scd_table }} where _airbyte_active_row = 0 and _ab_cdc_deleted_at is not null
```

could be inefficient (the partition `_airbyte_active_row = 0` is much larger than `_airbyte_active_row = 1` over time)

Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1117]: BTW would it work to do a pre-hook delete on the final table instead of a post-hook?

Isn't that easier to handle?

Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1117]: Did you try to run the deletes from the final table as a post hook of the scd table too?

File: airbyte-integrations/bases/base-normalization/dbt-project-template/macros/incremental.sql (hunk 3 above)
Comment (edgao) [reply airbyte-integrations/bases/base-normalization/dbt-project-template/macros/incremental.sql:7]: I did mess this up though :P the lack of syntax highlighting actually makes it pretty hard to edit this file

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py (hunk 2 above)
Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1117]: if you struggle with CTEs, you can make a separate materialized table (or view) instead of CTE (similar to the *_ab4 models that you can then use in the delete statement, etc)

**Review by edgao — COMMENTED** [review]
still wip, just wanted to back up my local changes. Running into vendor-specific test failures, so I need to work through those one by one.

**Comment by edgao** [comment]
still need to regenerate the test outputs, but I think this is finally ready for review. Integration tests are mostly working, just need to figure out clickhouse-specific syntax + how to run sql server on M1

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 6]
```diff
@@ -1106,20 +1137,112 @@ def add_to_outputs(
...
+                                from {{ quoted_scd_new_data_table }}
+                                where 1=1
+                                    {{ incremental_clause }}
+                            ),
+                            scd_active_rows as (
+                                select scd_table.* from {{ '{{ this }}' }} scd_table
+                                inner join modified_ids on scd_table.{{ unique_key }} = modified_ids.{{ unique_key }}
+                                where {{ active_row_column_name }} =  1
+                            )
+                            select modified_ids.{{ unique_key }} from scd_active_rows
```
Comment (edgao) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1201]: dbt-clickhouse is, for some reason, automatically injecting the schema name here (i.e. changing the CTE reference to `test_normalization.scd_active_rows`), so the query errors out because the table doesn't exist. Still looking into how to fix that. The other destinations all work fine though; will regenerate their normalization outputs later today.

File: airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/bigquery/test_nested_streams/dbt_project.yml [hunk 7]
```diff
@@ -57,7 +41,79 @@ models:
...
 dispatch:
   - macro_namespace: dbt_utils
-    search_order: ["airbyte_utils", "dbt_utils"]
+    search_order:
+      - airbyte_utils
+      - dbt_utils
+vars:
+  json_column: _airbyte_data
+  models_to_source:
+    nested_stream_with_complex_columns_resulting_into_long_names_ab1: test_normalization._airbyte_raw_nested_stream_with_complex_columns_resulting_into_long_names
```
[the remark this answers is not in the document]
Comment (edgao) [reply airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/bigquery/test_nested_streams/dbt_project.yml:52]: not entirely sure, but I think this has been happening for a while (and we just haven't regenerated the test outputs in a long time). Some files already have this, e.g. snowflake https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/snowflake/test_simple_streams/dbt_project.yml#L52

**Review by tuliren — APPROVED** [review]
Wow, quite a PR!

Some of the jinja template construction can probably be done with Python string interpolation to promote readability. But not a big deal.

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 8]
```diff
@@ -819,18 +858,9 @@ def generate_scd_type_2_model(self, from_table: str, column_names: Dict[str, Tup
             jinja_variables["scd_columns_sql"] = scd_columns_sql
         sql = Template(
             """
--- depends_on: {{ from_table }}
```
[the remark this answers is not in the document]
Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:850]: yes it's mostly for helping whoever that need to read the generated SQL to try to tweak it, it's still an auto-generated comment.

Maybe in some cases, dbt needs the hint but maybe here it's not very useful to it

**Comment by ChristopheDuong** [comment]
This is indeed a tricky PR... and not easy to review/verify by just reading it :(

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 9]
```diff
@@ -1106,20 +1137,123 @@ def add_to_outputs(
...
+                                                {{ primary_key }},
+                                            {%- endfor %}
+                                        ]) {{ '}}' }} as {{ unique_key }}
+                                    from {{ quoted_scd_new_data_table }}
+                                    where 1=1
+                                        {{ incremental_clause }}
+                                ) modified_ids
+                                on modified_ids.{{ unique_key }} = scd_active_rows.{{ unique_key }}
+                            group by modified_ids.{{ unique_key }}
+                            having count(scd_active_rows.{{ unique_key }}) = 0
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1216]: Can this be simplified??

I would feel more comfortable if we could do a different `incremental_clause` than a complex join to identify what needs to be dropped?


For example, can we maybe use the fact that we have a `normalized_at` column and identify all the freshly written rows from scd tables (they have been written after `max(normalized_at)` column of the final table)

We don't have a reason to stick to `airbyte_emitted_at` columns only?

Comment (edgao) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1216]: lol. I was actually trying to avoid that - I feel like incremental logic is already complicated enough, and being incremental against two columns which function similarly but not identically seemed really confusing. I'll try spiking it out, maybe it'll turn out to be really straightforward

Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1216]: that sounds much simpler!
is this still a post-hook on the scd table?

Comment (edgao) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1216]: yeah. I think ideally it would be a pre-hook on the final table, but suspect that might cause some weird behavior with repeated failures

Comment (edgao) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py]: pushed some code just for local backup. will run a few local syncs for further testing and rerequest review when I'm confident it works

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 10]
```diff
@@ -1154,17 +1158,24 @@ def add_to_outputs(
...
                         from {{ '{{ this }}' }}
-                        where {{ active_row_column_name }} = 1 {{ normalized_at_incremental_clause }}
+                        left join (
+                            select {{ unique_key }} as active_unique_key
+                            from {{ '{{ this }}' }}
+                            where {{ active_row_column_name }} = 1 {{ normalized_at_incremental_clause }}
+                        ) active_recent_scd_rows on {{ unique_key }} = active_unique_key
+                        where 1=1 {{ normalized_at_incremental_clause }}
+                        group by {{ unique_key }}
+                        having count(active_unique_key) = 0
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:1177]: is it mandatory to use of `having` clause? I'm not sure about performance if it's not less efficient than other approaches using `where` if possible instead?

**Review by ChristopheDuong — APPROVED** [review]

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py (hunk 10 above)
Comment (edgao) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py]: running some manual performance tests; at first glance neither version has been significantly faster, but the non-having version does at least feel more readable :shrug:

**Comment by edgao** [comment]
ok, one last attempt at test runs - failure seems to be from someone clearing out the integrationtests DB while the tests were running.

**Comment by edgao** [comment]
`DestinationType.REDSHIFT-1000` failed with oom error as expected; everything else passed. publishing with run-tests=false.
